### PR TITLE
Nightly pipeline

### PR DESCRIPTION
**Pull request by MaryGao**
fixes https://github.com/Azure/autorest.typescript/issues/1824

New nightly pipeline: https://dev.azure.com/azure-sdk/public/_build?definitionId=6134&_a=summary

### CODE REVIEW

File: .scripts/nightly.yml [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    displayName: "Re-build after update the config"
+  - script: npm list
+    workingDirectory: $(Build.SourcesDirectory)/packages/typespec-ts
+    displayName: "List the current dependency"
+  - script: npm run unit-test
+    workingDirectory: $(Build.SourcesDirectory)/packages/typespec-ts
+    displayName: "Run all unit test cases"
+  - script: npm run integration-test-ci
+    workingDirectory: $(Build.SourcesDirectory)/packages/typespec-ts
+    displayName: "Run all Cadl-ranch test cases"
```
Comment (qiaozha) [inline .scripts/nightly.yml:29]: why not run smoke-test in typespec-test folder

File: .scripts/nightly.yml [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+        - main
+
+steps:
+  - script: npm install -g @typespec/compiler@next
+    displayName: Install @typespec/compiler@next
+  - template: build.yml
+  - script: npm run override-dev-version
+    workingDirectory: $(Build.SourcesDirectory)/packages/typespec-ts
+    displayName: "Override pnpm-config version to latest dev"
+  - script: npx @azure-tools/typespec-bump-deps package.json
```
Comment (qiaozha) [inline .scripts/nightly.yml:15]: Not very clear why we still need this if we have override-dev-version script ?

File: packages/typespec-ts/test/commands/override-dev-version.ts [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    "The following is a mapping between packages and the versions we will update them to"
+  );
+  // eslint-disable-next-line no-console
+  console.log(packageToVersionRecord);
+  const packageJsonPaths = process.argv.slice(2);
+  for (const packageJsonPath of packageJsonPaths) {
+    const content = await readFile(packageJsonPath);
+    console.log(`Original content type: ${content.toString()}`);
+    const packageJson = JSON.parse(content.toString());
+    const depTypes = ["globalOverrides"];
```
Comment (qiaozha) [inline packages/typespec-ts/test/commands/override-dev-version.ts:35]: not quite sure why we need this ? npm handle this differently ?

File: packages/typespec-ts/test/commands/override-dev-version.ts [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+  "@typespec/rest",
+  "@typespec/http",
+  "@typespec/versioning",
+  "@azure-tools/typespec-client-generator-core",
+  "@azure-tools/typespec-azure-core",
+  "@typespec/eslint-config-typespec"
+];
+
+async function getKnownPackageVersion(packageName: string): Promise<string> {
+  return (await pacote.manifest(`${packageName}@next`)).version;
```
Comment (qiaozha) [inline packages/typespec-ts/test/commands/override-dev-version.ts:15]: can we use @next in the package.json directly ?

File: .scripts/nightly.yml (hunk 1 above)
Comment (MaryGao) [reply .scripts/nightly.yml:29]: let me add that

File: .scripts/nightly.yml (hunk 2 above)
Comment (MaryGao) [reply .scripts/nightly.yml:15]: If we don't add this the command `npm list` would fail

**Review by qiaozha — APPROVED** [review]

**Pull request merged**